lib/Sema/CSDiag.cpp

```diff
@@ -4264,8 +4264,14 @@ bool TypeChecker::diagnoseSelfAssignment(const Expr *E) {
   if (!AE)
     return false;
 
+  // Allow the use of parentheses to express self-assignment when it's a deliberate choice
+  auto src = AE->getSrc();
+  auto loadSrc = dyn_cast<LoadExpr>(src);
```

Comment:
You shouldn't be looking for `LoadExpr` here. If the RHS is a `let`, there won't be a `LoadExpr` anywhere.